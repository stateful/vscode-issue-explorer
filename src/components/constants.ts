import type { ExtensionConfiguration, WebViewState } from '../types'

export const vscode = acquireVsCodeApi<WebViewState>()

const configDataProperty = 'data-extension-configuration'
export const config: ExtensionConfiguration = JSON.parse(
    document.querySelector(`meta[${configDataProperty}]`)?.getAttribute(configDataProperty) as string
)

const codiconStyleSheet: CSSStyleSheet = []
    .slice.call(document.styleSheets)
    .find((file: any) => file.href && file.href.endsWith('codicon.css'))!
export const codiconCSSRules: string[] = [].slice.call(codiconStyleSheet.cssRules).map((r: CSSFontFaceRule) => r.cssText)
