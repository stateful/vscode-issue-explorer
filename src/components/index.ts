export { ExtensionTutorialComponent } from './tutorial'
export { IssueCreateForm } from './issueCreate'
