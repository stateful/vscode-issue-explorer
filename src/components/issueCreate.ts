import {
    html,
    css,
    unsafeCSS,
    CSSResult,
    LitElement,
    customElement
} from 'lit-element'
import { when } from 'lit/directives/when.js'
import Channel from 'tangle/webviews'
import shrinkPath from 'shrink-path'
import type { TextArea, TextField } from '@vscode/webview-ui-toolkit'
import type { Client } from 'tangle'

import { vscode, config, codiconCSSRules } from './constants'
import { ISSUE_CREATE_CHANNEL } from '../constants'
import type { CodeSelection, CreateIssueResult, CreateIssueResponse, CreateIssueError } from '../types'
import type { WebviewEvents } from '../webviews/issueCreate'

const MARKETPLACE_URL = 'https://marketplace.visualstudio.com/items?itemName=stateful.marquee'

@customElement('issue-create')
export class IssueCreateForm extends LitElement {
    #error?: string
    #result?: CreateIssueResult
    #client: Client<WebviewEvents>
    #codeSelection: CodeSelection[] = []
    #requestPending = false

    static get styles(): CSSResult[] {
        return [css/*css*/`
        vscode-radio-group {
            margin: 10px 0
        }

        a {
            color: #F52558
        }

        sub {
            display: block;
            margin-top: 5px;
        }

        img, vscode-text-field, vscode-text-area {
            width: 100%
        }

        .btnSection {
            display: flex;
            flex-direction: row;
            justify-content: flex-end;
            gap: 10px;
        }

        .codeSelection {
            padding-inline-start: 25px;
        }

        .codeSelection li {
            margin-bottom: 15px;
        }

        .errorMessage {
            color: red;
            font-weight: bold;
        }

        .codicon-check {
            color: green;
        }
        .codicon-check,
        .codicon-error {
            position: relative;
            top: 3px;
            padding-right: 5px;
        }

        .createIssueForm {
            opacity: 1;
            transition: opacity .2s linear;
        }
        .createIssueForm.hidden {
            opacity: 0;
        }

        @keyframes spin {
            from {
                transform: rotate(0deg);
            } to {
                transform: rotate(360deg);
            }
        }

        .codicon-loading {
            animation-name: spin;
            animation-duration: 4000ms;
            animation-iteration-count: infinite;
            animation-timing-function: linear;
        }
        `,
        ...codiconCSSRules.map((r: string) => unsafeCSS(r))]
    }

    constructor() {
        super()

        const channel = new Channel<WebviewEvents>(ISSUE_CREATE_CHANNEL)
        this.#client = channel.attach(vscode as any)
        this.#client.on('initIssueForm', this.#initIssueForm.bind(this))
        this.#client.on('issueCreateResult', this.#renderResult.bind(this))
    }

    render() {
        if (!this.#codeSelection || this.#codeSelection.length === 0) {
            return this.renderWelcomeView()
        }

        return html/* html */`
        ${when(this.#error, () => this.renderError(this.#error!))}
        ${when(this.#result,
            () => this.renderSuccess(this.#result!),
            () => this.renderForm()
        )}
        `
    }

    #cancel () {
        this.#requestPending = false
        this.#codeSelection = []
        this.requestUpdate()
    }

    #submit () {
        this.#requestPending = true
        this.requestUpdate()

        const titleElem = this.shadowRoot?.querySelector('vscode-text-field') as TextField
        const descriptionElem = this.shadowRoot?.querySelector('vscode-text-area') as TextArea
        this.#client.emit('issueCreateSubmission', {
            title: titleElem.value,
            description: descriptionElem.value,
            selection: this.#codeSelection
        })
    }

    renderWelcomeView() {
        return html/* html */`
        <h3>GitHub Issue Explorer</h3>
        <p>
            The <a href="${MARKETPLACE_URL}">GitHub Issue Explorer</a>
            allows you to quickly create GitHub issues while working on
            your code. Just select a line or multiple lines, do right
            click and pick <i>Create Issue From Selection</i>, e.g.:
        </p>
        <img src="${config.baseAppUri}/assets/img/demo.gif" />
        <p>
            Then fill out the issue form, press submit and done!
        </p>
        `
    }

    renderForm () {
        return html/*html*/`<form class="createIssueForm">
            <h3>Create Code Issue</h3>
            <p>
                <vscode-text-field name="title">Issue Title</vscode-text-field>
            </p>
            <p>
                <vscode-text-area
                    name="description"
                    rows=${10}
                >
                    Issue Description
                </vscode-text-area>
            </p>
            <p class="btnSection">
                <vscode-button @click=${this.#cancel.bind(this)} appearance="secondary">Cancel</vscode-button>
                <vscode-button @click=${this.#submit.bind(this)} ?disabled=${this.#requestPending}>
                    Submit
                    ${this.#requestPending && html`<span slot="start" class="codicon codicon-loading"></span>` || ''}
                </vscode-button>
            </p>
            <vscode-divider></vscode-divider>
            <h4>Selected Code Lines</h4>
            <ul class="codeSelection">
                ${this.#codeSelection.map((selection) => html/*html*/`
                <li>
                    ${shrinkPath(selection.uri, 50)}
                    <sub>${
                        selection.start === selection.end
                        ? html/*html*/`<b>Line:</b> ${selection.start + 1}</sub>`
                        : html/*html*/`<b>Lines:</b> ${selection.start + 1} - ${selection.end + 1}</sub>`
                    }
                </li>
                `)}
            </ul>
        </form>`
    }

    renderSuccess (result: CreateIssueResult) {
        return html/*html*/`
        <p>
            <span class='codicon codicon-check'></span>
            Issue successfully created ðŸŽ‰
        </p>
        <p>
            <b><a href="${result.issueUrl}">${result.issueLabel}</a></b>
            <sub>by <b><a href="${result.authorUrl}">${result.authorLabel}</a></b></sub>
        </p>
        `
    }

    renderError (error: string) {
        return html/*html*/`<p class="errorMessage">
            <span class='codicon codicon-error'></span>
            <b>${error}</b>
        </p>`
    }

    #initIssueForm(codeSelection: CodeSelection[]) {
        this.#error = undefined
        this.#result = undefined
        this.#requestPending = false
        this.#codeSelection = codeSelection
        this.requestUpdate()
    }

    async #renderResult(result: CreateIssueResponse) {
        this.#requestPending = false

        const { error } = result as CreateIssueError
        if (error) {
            this.#error = error
            return this.requestUpdate()
        }

        const form = this.shadowRoot?.querySelector('.createIssueForm')
        if (!form) {
            return
        }

        form.className += ' hidden'
        await new Promise((resolve) => setTimeout(resolve, 200))
        form.remove()
        this.#error = undefined
        this.#result = result as CreateIssueResult
        return this.requestUpdate()
    }
}
