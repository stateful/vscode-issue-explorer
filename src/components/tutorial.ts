import {
    html,
    css,
    state,
    CSSResult,
    property,
    LitElement,
    customElement
} from 'lit-element'
import { when } from 'lit/directives/when.js'
import { config } from './constants'

@customElement('extension-tutorial')
export class ExtensionTutorialComponent extends LitElement {
    @property({ type: String })
    baseUri: string = ''

    @state()
    protected _isOpen = false

    static get styles(): CSSResult {
        return css/*css*/`
        .tutorial {
            transition: opacity .2s ease-out;
            display: none;
            opacity: 0;
            margin-bottom: 15px;
        }
        .tutorial.open {
            display: block;
            opacity: 1;
        }
        .footer-link:before {
            content: '→';
        }
        .footer-link.open:before {
            content: '✗';
        }
        img {
            width: 100%;
        }
        `
    }

    render () {
        const linkLabel = this._isOpen ? 'Close Tutorial' : 'Open Tutorial'
        return html/*html*/`
        <p>
            <vscode-link
                class="footer-link${when(this._isOpen, () => ' open')}"
                @click=${this.toggleTutorial.bind(this)}
            >
                ${linkLabel}
            </vscode-link>
        </p>
        <div class="tutorial${when(this._isOpen, () => ' open')}">
            <img src="${config.baseAppUri}/assets/img/select.gif" style="max-width: 594px" />
            <p>
                Then fill out the issue form and connect for code references if desired.
            </p>
            <img src="${config.baseAppUri}/assets/img/describe.gif" style="max-width: 269px" />
            <p>
                You can comment on individual code references to document
                what this piece of code is about and what changes you'ld expect.
            </p>
            <img src="${config.baseAppUri}/assets/img/details.gif" style="max-width: 269px" />
            <p>
                Lastly, click on submit and review your issue on GitHub:
            </p>
            <img src="${config.baseAppUri}/assets/img/view.gif" style="max-width: 717px" />
            <p>
                Or explore created issues within VS Code:
            </p>
            <img src="${config.baseAppUri}/assets/img/explore.gif" style="max-width: 671px" />
        </div>
        `
    }

    toggleTutorial () {
        this._isOpen = !this._isOpen
    }
}
