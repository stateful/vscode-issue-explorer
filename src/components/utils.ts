import type { Client } from 'tangle'

import { vscode } from './constants'
import type { WebViewState, WebviewEvents } from '../types'

export const updateState = (client: Client<WebviewEvents>, stateChange: Partial<WebViewState>) => {
    const currentState: WebViewState = vscode.getState() || <WebViewState>{}
    const newState: WebViewState = { ...currentState, ...stateChange }
    vscode.setState(newState)
    return client.emit('stateUpdate', newState)
}
