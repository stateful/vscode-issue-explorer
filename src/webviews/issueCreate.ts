import {
    Webview,
    WebviewView,
    WebviewViewProvider,
    ExtensionContext
} from 'vscode'

import { webviewOptions } from '../constants'
import { getHtmlForWebview } from '../utils'

export default class IssueCreate implements WebviewViewProvider {
    constructor(private readonly _context: ExtensionContext) {}

    async resolveWebviewView(webviewView: WebviewView): Promise<void> {
        webviewView.webview.html = await getHtmlForWebview(webviewView.webview, this._context.extensionUri)
        webviewView.webview.options = {
            ...webviewOptions,
            localResourceRoots: [this._context.extensionUri],
        }
        console.log('[IssueCreate] webview resolved')
    }
}
