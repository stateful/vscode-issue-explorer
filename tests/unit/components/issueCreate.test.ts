/**
 * @vitest-environment jsdom
 */
import { EventEmitter } from 'events'
import { test, expect, vi, beforeEach } from 'vitest'
import { fixture, assert } from '@open-wc/testing'
import { html } from 'lit'
import type { Link, TextField, TextArea, Button } from '@vscode/webview-ui-toolkit'
// @ts-expect-error mock feature
import { emitter } from 'tangle/webviews'

import { IssueCreateForm } from '../../../src/components/issueCreate'

vi.mock('tangle/webviews', () => {
    let emitter = new EventEmitter()
    const origEmit = emitter.emit.bind(emitter)
    emitter.emit = vi.fn().mockImplementation(origEmit)
    return {
        emitter,
        default: class TangleMock {
            public attach = vi.fn().mockReturnValue(emitter)
        }
    }
})

vi.mock('../../../src/components/constants', () => ({
    vscode: {
        getState: vi.fn().mockReturnValue({ foo: 'bar' }),
        setState: vi.fn()
    },
    config: {
        baseAppUri: '/foo/bar'
    },
    codiconCSSRules: []
}))

const EXAMPLE_SELECTION = {
    start: 1,
    end: 2,
    uri: '/foo/bar',
    fileType: 'js',
    code: 'console.log("foobar")'
}

const EXAMPLE_RESULT = {
    issueUrl: '/url/to/issue',
    issueLabel: 'some title',
    authorUrl: '/url/to/author',
    authorLabel: 'some author'
}

beforeEach(() => {
    vi.mocked(emitter.emit).mockClear()
})

test('is defined', () => {
    const el = document.createElement('issue-create')
    assert.instanceOf(el, IssueCreateForm)
})

test('renders as default the intro screen', async () => {
    const el = await fixture(html`<issue-create></issue-create>`)
    expect(el.shadowRoot!.innerHTML).toContain('<h2>GitHub Issue Explorer</h2>')
})

test('should switch to form once a selection was made', async () => {
    const el: IssueCreateForm = await fixture(html`<issue-create></issue-create>`)
    emitter.emit('initIssueForm', [EXAMPLE_SELECTION, EXAMPLE_SELECTION])
    await el.updateComplete
    expect(el.shadowRoot!.innerHTML).toContain('<h2>Create Code Issue</h2>')
    expect(el.shadowRoot!.querySelectorAll('.codeSelection li').length).toBe(2)

    const codeContent = el.shadowRoot!.querySelector('.codeSelection li')!.textContent?.trim()
    expect(codeContent).toContain('/foo/bar')
    expect(codeContent).toContain('Lines: 2 - 3')
})

test('cancel should lead back to start screen', async () => {
    const el: IssueCreateForm = await fixture(html`<issue-create></issue-create>`)
    emitter.emit('initIssueForm', [EXAMPLE_SELECTION, EXAMPLE_SELECTION])
    await el.updateComplete
    expect(el.shadowRoot!.innerHTML).toContain('<h2>Create Code Issue</h2>')
    ;(el.shadowRoot!.querySelector('vscode-button[appearance="secondary"]') as Link).click()
    await el.updateComplete
    expect(el.shadowRoot!.innerHTML).toContain('<h2>GitHub Issue Explorer</h2>')
})

test('renders error', async () => {
    const el: IssueCreateForm = await fixture(html`<issue-create></issue-create>`)
    emitter.emit('initIssueForm', [EXAMPLE_SELECTION, EXAMPLE_SELECTION])
    emitter.emit('issueCreateResult', { error: 'ups!' })
    await el.updateComplete
    expect(el.shadowRoot!.querySelector('.errorMessage b')!.textContent).toBe('ups!')
})

test('can remove code reference', async () => {
    const el: IssueCreateForm = await fixture(html`<issue-create></issue-create>`)
    emitter.emit('initIssueForm', [EXAMPLE_SELECTION, EXAMPLE_SELECTION])
    await el.updateComplete
    expect(el.shadowRoot!.querySelectorAll('.codeSelection li').length).toBe(2)
    ;(el.shadowRoot!.querySelector('button[title="Remove Code Reference"]') as Link).click()
    await el.updateComplete
    expect(el.shadowRoot!.querySelectorAll('.codeSelection li').length).toBe(1)
})

test('shows success message', async () => {
    const el: IssueCreateForm = await fixture(html`<issue-create></issue-create>`)
    emitter.emit('initIssueForm', [EXAMPLE_SELECTION, EXAMPLE_SELECTION])
    await el.updateComplete
    emitter.emit('issueCreateResult', EXAMPLE_RESULT)
    await new Promise((resolve) => setTimeout(resolve, 200))
    await el.updateComplete
    expect(el.shadowRoot!.innerHTML).toContain('Issue successfully created')
    expect([...el.shadowRoot!.querySelectorAll('vscode-link')].map((l) => l.textContent))
        .toEqual(['some title', 'some author'])
})

test('should be able to submit something', async () => {
    const el: IssueCreateForm = await fixture(html`<issue-create></issue-create>`)
    emitter.emit('initIssueForm', [EXAMPLE_SELECTION, EXAMPLE_SELECTION])
    await el.updateComplete
    ;([...el.shadowRoot!.querySelectorAll('vscode-text-area')]).forEach(
        ((elem: TextField, i) => { elem.value = `some text ${i}`}) as any)
    ;(el.shadowRoot!.querySelector('vscode-text-field') as TextArea).value = 'some long description'
    ;(el.shadowRoot!.querySelector('button[title="Add Comment to Code Reference"') as Button).click()
    ;(el.shadowRoot!.querySelectorAll('.btnSection vscode-button')[1] as Button).click()
    await el.updateComplete
    expect(emitter.emit.mock.calls.pop()).toMatchSnapshot()
})
