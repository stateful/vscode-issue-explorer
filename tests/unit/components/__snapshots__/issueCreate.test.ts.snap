// Vitest Snapshot v1

exports[`should be able to submit something 1`] = `
[
  "issueCreateSubmission",
  {
    "description": "some text 0",
    "selection": [
      {
        "code": "console.log(\\"foobar\\")",
        "comment": "",
        "end": 2,
        "fileType": "js",
        "start": 1,
        "uri": "/foo/bar",
      },
      {
        "code": "console.log(\\"foobar\\")",
        "comment": "",
        "end": 2,
        "fileType": "js",
        "start": 1,
        "uri": "/foo/bar",
      },
    ],
    "title": "some long description",
  },
]
`;
